### PR TITLE
[Enhancement] Unify Random Forest and Extra Trees preview algorithms

### PR DESCRIPTION
**Pull request by icfaust**
# Description
This unifies Random Forest and Extra Trees to a single file from the same inheritance classes to match stock scikit-learn. This reduces overall code size, and prepares for movement out of preview. This improves preview RF to meet some of the standards of sklearnex in terms of testing and verbosity. It is still an open question whether to rename extra_trees.py to forest.py, though this would be temporary before the move out of preview.

Changes proposed in this pull request:
- Fix TODO associated with _estimator base_estimator_ and estimator_
- add sklearn v1.4 monotonic_cst keyword and API input
- move RF algos to extra_trees.py, rename to _forest.py (matches sklearn/daal4py naming convention)
- remove forest.py
- subtle reorder of patching_status to exactly match daal4py RF order
- correct issue associated with 'balanced_subsample' in class_weight
- remove _onedal_model attribute and focus on _onedal_estimator instead
- fix testing issue with test_min_samples_leaf for daal4py
- spoof inheritance of RF/ET algos to allow for proper isinstance checking
- fix fallback issue for warm_start caused by estimators_ lazy evaluation
- Re-add CI tests from deselected_tests for RF and ET
- splitter_mode kwarg in preview RF is removed, splitter mode is checked by the passed tree object to match sklearn
- Change import in RF spmd interface (generalization)
- NOTE: binning strategy (see daal4py RF) not implemented

### CODE REVIEW

File: sklearnex/preview/ensemble/__init__.py [hunk 1]
```diff
@@ -15,8 +15,12 @@
...
 # ===============================================================================
 
-from .extra_trees import ExtraTreesClassifier, ExtraTreesRegressor
-from .forest import RandomForestClassifier, RandomForestRegressor
+from .extra_trees import (
+    ExtraTreesClassifier,
+    ExtraTreesRegressor,
+    RandomForestClassifier,
+    RandomForestRegressor,
+)
```
Comment (Alexsandruss) [inline sklearnex/preview/ensemble/__init__.py:23]: `extra_trees` module should be renamed to a common name for ExtraTrees and RandomForest

**Comment by samir-nasibli** [comment]
@icfaust could you please also check what RF tests will enabled with this update?

**Comment by samir-nasibli** [comment]
@icfaust please move `sklearnex/preview/ensemble/extra_trees.py → sklearnex/preview/ensemble/_forest.py`

**Comment by icfaust** [comment]
> @icfaust please move `sklearnex/preview/ensemble/extra_trees.py → sklearnex/preview/ensemble/_forest.py`

See https://github.com/intel/scikit-learn-intelex/pull/1440/commits/c34283060d28752f8a1428aff4eaa3b81c317cec

**Comment by samir-nasibli** [comment]
> > @icfaust please move `sklearnex/preview/ensemble/extra_trees.py → sklearnex/preview/ensemble/_forest.py`
> 
> See [c342830](https://github.com/intel/scikit-learn-intelex/commit/c34283060d28752f8a1428aff4eaa3b81c317cec)

Please be careful, it seems you moved to the root of the repository :)

**Comment by icfaust** [comment]
NOTE: THIS PR IS NOW BREAKING TESTING BY THE LAST COMMIT ON PURPOSE AND MUST BE REVERTED BEFORE REVIEW.

**Comment by icfaust** [comment]
GPU testing with deselected_tests pullback http://intel-ci.intel.com/ee4d808d-89e0-f1d4-a9d7-a4bf010d0e2e

File: daal4py/sklearn/ensemble/_forest.py [hunk 2]
```diff
@@ -1355,7 +1357,9 @@ def _daal_fit_regressor(self, X, y, sample_weight=None):
         if isinstance(self.min_samples_leaf, numbers.Integral):
             parameters["minObservationsInLeafNode"] = self.min_samples_leaf
         else:
-            int(ceil(self.min_samples_leaf * X.shape[0]))
+            parameters["minObservationsInLeafNode"] = int(
+                ceil(self.min_samples_leaf * X.shape[0])
+            )
```
Comment (KulikovNikita) [inline daal4py/sklearn/ensemble/_forest.py:1362]: I believe `ceil` returns an `Integral` value [by default](https://docs.python.org/3/library/math.html#math.ceil)

**Comment by icfaust** [comment]
Pre commit with GPU testing with readded deselected_tests.yaml http://intel-ci.intel.com/ee527ce8-e814-f1b2-abb8-a4bf010d0e2e

**Comment by icfaust** [comment]
Public CI is green. If pre-commit is green, I will fully remove the commented-out deselected_tests and rerun.

**Comment by icfaust** [comment]
Associated nightly run for deselected_test verification http://intel-ci.intel.com/ee52f85f-6140-f176-bdcf-a4bf010d0e2e

**Comment by icfaust** [comment]
> NOTE: THIS PR IS NOW BREAKING TESTING BY THE LAST COMMIT ON PURPOSE AND MUST BE REVERTED BEFORE REVIEW.

Applicable deselections have been now reapplied. Testing restored.

**Comment by icfaust** [comment]
Pre-commit issue is due to a bad mpi infra install in one case. It is unrelated to these changes, otherwise it is all green, including private CI nightly.

**Comment by icfaust** [comment]
Private nightly rerun with change http://intel-ci.intel.com/ee534500-d546-f1f5-9a78-a4bf010d0e2e

**Comment by icfaust** [comment]
> Private nightly rerun with change http://intel-ci.intel.com/ee534500-d546-f1f5-9a78-a4bf010d0e2e

This is green. There exists now a question about the RF spmd interface. @ethanglaser  @samir-nasibli, is there anything that preview might be doing to interact with the RF SPMD interface?

**Comment by icfaust** [comment]
New nightly run for spmd change: http://intel-ci.intel.com/ee539b0d-e065-f146-84b2-a4bf010d0e2e

**Comment by icfaust** [comment]
> New nightly run for spmd change: http://intel-ci.intel.com/ee539b0d-e065-f146-84b2-a4bf010d0e2e

Non-RF algos causing failures in this run.

**Comment by ethanglaser** [comment]
> > New nightly run for spmd change: http://intel-ci.intel.com/ee539b0d-e065-f146-84b2-a4bf010d0e2e
> 
> Non-RF algos causing failures in this run.

I see no issues with this job (possibly something sporadic that could be resolved with a re-run.

**Comment by icfaust** [comment]
@samir-nasibli @Alexsandruss @KulikovNikita Please review, thanks!

**Review by inteldimitrius — APPROVED** [review]
LGTM

**Comment by icfaust** [comment]
http://intel-ci.intel.com/ee5c8a0d-cac9-f1b7-839d-a4bf010d0e2e private ci rerun for infra issues

**Pull request merged**